### PR TITLE
NEXT-19865 - Use new twig template name for each unique template to remove statefulness

### PR DESCRIPTION
**Pull request by JoshuaBehrens**
### 1. Why is this change necessary?
When you have a broken SEO URL template it can silently break follow up usages.

### 2. What does this change do, exactly?
1. Add a logger so you can at least have a chance of spotting the error in the log files
2. Make a new instance of twig each time you want to generate URLs as you can't clone a twig environment safely
3. Do not try to generate URLs with a broken twig (reduces wall time and makes it faster to spot issues)
4. Remove a stateful service by having a twig environment for each generator

### 3. Describe each step to reproduce the issue or behaviour.
1. Have a broken twig template as SEO URL template
2. Run `dal:refresh:index` to re-index category URLs
3. Bake a cake on the computer in the meantime
4. Have no new URLs
5. Admin says green tick behind pattern
6. Check log files. Nothing there
7. Debug a lot
8. Find twig trying to render a category URL with a template of the categories but the compiled twig class for rendering product URLs?!?!?
9. Find muted exceptions
10. Curl up and crying oneself to sleep

### 4. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### 5. Extras

Big thanks is going to @remco-verton for his twig debugging and detect the cache poisoning from using the same name. This knowledge made this pull request non-breaking!

### CODE REVIEW

**Comment by JoshuaBehrens** [comment]
Yes, this conflicts with https://github.com/shopware/platform/pull/2087

File: src/Core/Content/Seo/SeoUrlGenerator.php [hunk 1]
```diff
@@ -26,31 +29,40 @@ class SeoUrlGenerator
...
+        $this->logger = $logger;
     }
 
-    /**
-     * @feature-deprecated (flag:FEATURE_NEXT_13410) Parameter $salesChannel will be required
-     */
     public function generate(array $ids, string $template, SeoUrlRouteInterface $route, Context $context, ?SalesChannelEntity $salesChannel): iterable
     {
         $criteria = new Criteria($ids);
-        $route->prepareCriteria($criteria, $salesChannel);
```
Comment (shyim) [inline src/Core/Content/Seo/SeoUrlGenerator.php]: revert this change

File: src/Core/Content/Seo/SeoUrlGenerator.php [hunk 2]
```diff
@@ -26,31 +29,40 @@ class SeoUrlGenerator
...
         $this->requestStack = $requestStack;
-        $this->twig = $environment;
+        $this->seoUrlTwigFactory = $seoUrlTwigFactory;
+        $this->slugify = $slugify;
+        $this->twigExtensions = $twigExtensions;
+        $this->logger = $logger;
     }
 
-    /**
-     * @feature-deprecated (flag:FEATURE_NEXT_13410) Parameter $salesChannel will be required
```
Comment (shyim) [inline src/Core/Content/Seo/SeoUrlGenerator.php]: same to this

File: src/Core/Content/Seo/SeoUrlGenerator.php [hunk 3]
```diff
@@ -63,19 +75,23 @@ public function generate(array $ids, string $template, SeoUrlRouteInterface $rou
             return new RepositoryIterator($repository, $context, $criteria);
         });
 
-        $this->setTwigTemplate($config, $template);
+        $twig = $this->createTwigTemplate($config, $template);
 
-        while ($entities = $iterator->fetch()) {
-            yield from $this->generateUrls($route, $config, $salesChannel, $entities);
+        if ($twig instanceof Environment) {
```
Comment (shyim) [inline src/Core/Content/Seo/SeoUrlGenerator.php]: why should be twig something different

File: src/Core/Content/Seo/SeoUrlGenerator.php [hunk 4]
```diff
@@ -63,19 +75,23 @@ public function generate(array $ids, string $template, SeoUrlRouteInterface $rou
...
-    {
-        $request = $this->requestStack->getMainRequest();
+    private function generateUrls(
+        Environment $twig,
+        SeoUrlRouteInterface $seoUrlRoute,
+        SeoUrlRouteConfig $config,
+        ?SalesChannelEntity $salesChannel,
+        EntityCollection $entities
+    ): iterable {
+        $request = $this->requestStack->getMasterRequest();
```
Comment (shyim) [inline src/Core/Content/Seo/SeoUrlGenerator.php]: this usage is deprecated

File: src/Core/Content/Seo/SeoUrlGenerator.php [hunk 5]
```diff
@@ -116,11 +132,15 @@ private function generateUrls(SeoUrlRouteInterface $seoUrlRoute, SeoUrlRouteConf
         }
     }
 
-    private function getSeoPathInfo(SeoUrlMapping $mapping, SeoUrlRouteConfig $config): ?string
+    private function getSeoPathInfo(Environment $twig, SeoUrlMapping $mapping, SeoUrlRouteConfig $config): ?string
     {
         try {
-            return trim($this->twig->render('template', $mapping->getSeoPathInfoContext()));
-        } catch (\Throwable $error) {
+            return trim($twig->render('template', $mapping->getSeoPathInfoContext()));
```
Comment (shyim) [inline src/Core/Content/Seo/SeoUrlGenerator.php]: does the twig caching still work?
are there any performance differences between using one instance and creating always one?

File: src/Core/Content/Seo/SeoUrlGenerator.php [hunk 6]
```diff
@@ -116,11 +132,15 @@ private function generateUrls(SeoUrlRouteInterface $seoUrlRoute, SeoUrlRouteConf
...
+        } catch (Error $error) {
+            $this->logger->critical('Error received on rendering SEO URL template', [
+                'exception' => $error,
```
Comment (shyim) [inline src/Core/Content/Seo/SeoUrlGenerator.php:139]: log the `$mapping`. Otherwise this will be pretty useless as you can't identifiy which seo url is failing here

File: src/Core/Content/Seo/SeoUrlGenerator.php [hunk 7]
```diff
@@ -129,18 +149,28 @@ private function getSeoPathInfo(SeoUrlMapping $mapping, SeoUrlRouteConfig $confi
...
         $template = "{% autoescape '" . self::ESCAPE_SLUGIFY . "' %}$template{% endautoescape %}";
-        $this->twig->setLoader(new ArrayLoader(['template' => $template]));
+        $twig->setLoader(new ArrayLoader(['template' => $template]));
 
         try {
-            $this->twig->loadTemplate($this->twig->getTemplateClass('template'), 'template');
+            $twig->loadTemplate($twig->getTemplateClass('template'), 'template');
         } catch (SyntaxError $syntaxError) {
+            $this->logger->critical('Error initializing SEO URL template', [
+                'exception' => $syntaxError,
```
Comment (shyim) [inline src/Core/Content/Seo/SeoUrlGenerator.php:167]: log the $mapping. Otherwise this will be pretty useless as you can't identifiy which seo url is failing here

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php [hunk 8]
```diff
@@ -194,6 +202,90 @@ public function testVariantInheritance(): void
...
+        }
+    }
+
+    public function testErrorLogging()
+    {
+        $logger = new class () extends AbstractLogger
+        {
+            public $logs = [];
+
+            public function log($level, $message, array $context = array())
```
Comment (shyim) [inline src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: can we typehint there maybe already for psr/log 3?

Comment (shyim) [reply src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: and isn't there any already existing logger for this?

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php [hunk 9]
```diff
@@ -194,6 +202,90 @@ public function testVariantInheritance(): void
...
+        $secondRun = $this->seoUrlGenerator->generate(
+            $categoryIds,
+            'template second run',
+            $this->seoUrlRouteRegistry->findByRouteName(TestNavigationSeoUrlRoute::ROUTE_NAME),
+            $this->salesChannelContext->getContext(),
+            $this->salesChannelContext->getSalesChannel()
+        );
+
+        /** @var SeoUrlEntity $url */
+        foreach ($firstRun as $url) {
```
Comment (shyim) [inline src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: why foreach, when you could get the first one from $firstRun

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php [hunk 10]
```diff
@@ -194,6 +202,90 @@ public function testVariantInheritance(): void
         static::assertEquals('parent/green', $urls[$ids->get('green')]);
     }
 
+    public function testNotBeingStateful(): void
+    {
+        $categoryIds = $this->getCategoryIds(2);
+
+        self::assertCount(2, $categoryIds, 'this is important for the test as you need more items to iterate for a context switch test');
```
Comment (shyim) [inline src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]:
```suggestion
        static::assertCount(2, $categoryIds, 'this is important for the test as you need more items to iterate for a context switch test');
```

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php [hunk 11]
```diff
@@ -194,6 +202,90 @@ public function testVariantInheritance(): void
...
+
+        /** @var \Generator $firstRun */
```
Comment (shyim) [inline src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: can you type it more explict what it is 😅

File: src/Core/Framework/DependencyInjection/seo.xml [hunk 12]
```diff
@@ -154,11 +157,5 @@
         </service>
 
         <service id="Shopware\Core\Content\Seo\SeoUrlTwigFactory"/>
-
-        <service id="shopware.seo_url.twig" class="Twig\Environment">
```
Comment (shyim) [inline src/Core/Framework/DependencyInjection/seo.xml]: this is an BC removing services

Comment (JoshuaBehrens) [reply src/Core/Framework/DependencyInjection/seo.xml]: I know. Did you read my changelog?

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php (hunk 11 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: Do I have to? It just needs to be a generator (what it really is) to be a traversable for later test usage.

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php (hunk 10 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: Already read that from the logs. Will amend #soon

File: src/Core/Framework/DependencyInjection/seo.xml (hunk 12 above)
Comment (shyim) [reply src/Core/Framework/DependencyInjection/seo.xml]: And? it doesn't matter. you can't break anything in trunk.

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php (hunk 9 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: I need to trigger an iteration of the generator. next would be sufficient as well but eventually same result.

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php (hunk 8 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: Sure I can.

I don't know. I just need to capture some logs to compare them later. Tell me :D Do you have such test tooling?

File: src/Core/Content/Seo/SeoUrlGenerator.php (hunk 7 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php:167]: Good hint!

File: src/Core/Content/Seo/SeoUrlGenerator.php (hunk 6 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php:139]: Good hint!

File: src/Core/Content/Seo/SeoUrlGenerator.php (hunk 5 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php]: The factory disables the twig cache anyway.

Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php]: Yes there is. First, you don't create the first twig environment in the container anymore. Second, you create an environment for each indexing call so either in a web requests which is similar to have it in the container or per indexing message in the message worker which is neglectable in impact IMO.

File: src/Core/Content/Seo/SeoUrlGenerator.php (hunk 4 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php]: I need to lookup where I got that one from. Maybe from my original local fix for shopware 6.3. When this usage is deprecated, what is the new fancy way?

File: src/Core/Content/Seo/SeoUrlGenerator.php (hunk 3 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php]: Read the factory code/type hint. It can be null when setting of the template fails.

File: src/Core/Content/Seo/SeoUrlGenerator.php (hunk 2 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php]: Yes, sorry. That is from my local shopware 6.3 fix.

File: src/Core/Content/Seo/SeoUrlGenerator.php (hunk 1 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php]: Yes, sorry. That is from my local shopware 6.3 fix.

File: src/Core/Framework/DependencyInjection/seo.xml (hunk 12 above)
Comment (JoshuaBehrens) [reply src/Core/Framework/DependencyInjection/seo.xml]: Oh, sorry. I thought it is your task to merge it so you don't break stuff. Ok ok ok. So which branch will I choose instead?

**Comment by shyim** [comment]
Btw why you can't just clone the twig instance?

**Comment by JoshuaBehrens** [comment]
I can't trust a clone of twig as it does not provide a clone implementation but contains arrays of e.g. extension instances which probably needs to be deep cloned. I don't want to introduce deep cloning here. Most deep cloning implementations are flawed by `spl_object_hash`.

**Comment by shyim** [comment]
I am just searching for a compatible way of the current services. Maybe https://symfony.com/doc/current/service_container/shared.html can help us here to ensure its always a new instance 🤔

**Comment by JoshuaBehrens** [comment]
@shyim I also thought about this feature but this only creates a new instance each time we define a new service. So I need to create a service for each template 🙃

File: src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php (hunk 8 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php]: Wait, we can't go psr/log 3 because this is php 8.0 only. Or is shopware already 8.0 only?

File: src/Core/Content/Seo/SeoUrlGenerator.php (hunk 4 above)
Comment (JoshuaBehrens) [reply src/Core/Content/Seo/SeoUrlGenerator.php]: Ah I see master vs main. Fixed

**Comment by JoshuaBehrens** [comment]
Big thanks is going to @remco-verton for his twig debugging and detect the cache poisoning from using the same name. This knowledge made this pull request non-breaking!

**Comment by celha** [comment]
@JoshuaBehrens can you please fix the failing tests from the PHPUnit Pipeline? It would be necessary for us that the tests are passing before we can merge this PR :)

**Comment by celha** [comment]
@JoshuaBehrens sorry for the late reply - I just saw that your PR still has some failing tests (PHPUnit & PHP Analyse). Please make sure that the tests are passing before we can merge your changes :)

**Comment by celha** [comment]
@JoshuaBehrens sorry for the late reply - I just saw your PR and there are still some failing tests (PHPUnit & PHP Analyse). Please make sure to make them pass, so that we can merge your changes :)

**Comment by celha** [comment]
@JoshuaBehrens sorry again for the late reply :( unfortunately, this has really gone down with us.. Would you be so kind and rebase this PR, so that we can merge it soon? :)

File: changelog/_unreleased/2022-01-22-fix-statefulness-issues-of-twig-environment-in-seo-url-generator.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
+---
+title: Fix issues by statefulness twig environment in SeoUrlGenerator
+author: JoshuaBehrens
+author_email: code@joshua-behrens.de
+author_github: @JoshuaBehrens
+---
+# Core
+* Fixed usage of broken twig templates in `Shopware\Core\Content\Seo\SeoUrlGenerator` to bullet-proof it against invalid configured SEO URL templates
```
Comment (marcelbrode) [inline changelog/_unreleased/2022-01-22-fix-statefulness-issues-of-twig-environment-in-seo-url-generator.md]: Please only use "Added", "Changed", "Removed" or "Deprecated" in changelogs

**Comment by marcelbrode** [comment]
Still waiting for the rebase :smile:

**Comment by JoshuaBehrens** [comment]
Gomen! I put it on my todo :D

**Comment by marcelbrode** [comment]
Any updates here? :slightly_smiling_face:

**Comment by JoshuaBehrens** [comment]
@marcelbrode rebased, change according to the "new" SeoExceptions, changed changelog :) have fun

**Comment by marcelbrode** [comment]
The PR has been merged. Thanks for your contribution :tada:

**Comment by JoshuaBehrens** [comment]
Thank you for merging :)
